Fix camera calibrator

Brow71189 opened the pull request:
This is what is needed to make dynamic calibrations work properly for the ELA together with correct fallback logic so that a software update does not need a config file update.

I split the changes into two commits because they are not dependent on each other.

I'm still not 100% happy with this, as it uses keys exisiting as mechanism to figure out what the user intends. So maybe we need to re-think the whole "camera_type", "signal_type" and calibration mechanism at some point. Currently, the camera needs to track controls in AS2 in order to figure out what the current settings are and therefore supply the correct calibrations. I know we would like to keep direct AS2 access out of the camera, but then we need a generalized mechanism of letting the camera know what signal it currently receives.
ALternatively we could move creation of calibrations and tracking of signal type out of the camera entirely. That might give us a cleaner separation but will be quite a bit of work.

cmeyer commented:
This PR should be split into two parts - one to use signal type to determine how to handle offset; and the other (possibly) to add functionality for backwards compatibility with camera devices.

I added a test that succeeds for the "old style" tests with the old style calibrations, which were last correct in 0e4d1705ac74327f5eb0ddca4f4f251ce69d030f.

Then I fixed the regression introduced in e027d481c94aeea944254477b8b6874cbd141f69 and added "new style" test sections to test the new functionality.

The first PR to use signal type to handle offsets should include an updated test or new test to test that functionality.

If you decide you want to proceed with the additional "x_offset_..." style functionality, it will need to pass the existing test and add a new test also. I'm not sure we need this new functionality, so let's discuss offline before proceeding.

cmeyer commented:
A new commit which adds `signal_type` handling f748f8215f604497b9b1b050da25e25edb629c93

Brow71189 commented:
> ~This PR should be split into two parts - one to use signal type to determine how to handle offset; and the other (possibly) to add functionality for backwards compatibility with camera devices.~
> 
> I added a test that succeeds for the "old style" tests with the old style calibrations, which were last correct in [0e4d170](https://github.com/nion-software/nionswift-instrumentation-kit/commit/0e4d1705ac74327f5eb0ddca4f4f251ce69d030f).
> 
> Then I fixed the regression introduced in [e027d48](https://github.com/nion-software/nionswift-instrumentation-kit/commit/e027d481c94aeea944254477b8b6874cbd141f69) and added "new style" test sections to test the new functionality in this commit [c2da92b](https://github.com/nion-software/nionswift-instrumentation-kit/commit/c2da92b3667ae88214eac7229ab3ce202959363c) (similar to your changes in this PR).
> 
> ~The first PR to use signal type to handle offsets should include an updated test or new test to test that functionality.~
> 
> If you decide you want to proceed with the additional "x_offset_..." style functionality, it will need to pass the existing test and add a new test also. I'm not sure we need this new functionality, so let's discuss offline before proceeding.
> 
> EDIT: I also added the `signal_type` functionality + a test for it here [f748f82](https://github.com/nion-software/nionswift-instrumentation-kit/commit/f748f8215f604497b9b1b050da25e25edb629c93)

Thanks for looking into this. I think the "signal_type" functionality is fine now.

The fallback logic for the old style calibration dict ("x_offset_...") is still not correct though. I don't understand why you consider the same names that are used in the new style dict as old style when you use them lowercase. Is there any place we actually use that? The main reason for this PR was to get the fallback logic fixed, otherwise we need to manually update all camera config files when we update the microscope software, which is not something we want.

cmeyer commented:
After discussion:
- the "tv_x_offset" style was never used in config files, it was only used as dict keys when controls were specified by the camera
- the lowercase key style was an unfortunate choice in the first pass, not sure if it was ever used, but maintaining backwards compatibility just in case
- the new camel case style is supported and now has a test and is sufficient for our needs
- the Dectris camera can use the new style config
- future plans may include hoisting all of this config out of the cameras themselves and into a common config file for all devices
